STYLEGUIDE.md

```diff
@@ -3,25 +3,695 @@
+
 *Right:*
-  - `src/kibana/index_patterns/index_pattern.js`
+```
+POST /api/kibana/index_patterns
+{
+  "id": "...",
+  "time_field_name": "...",
+  "fields": [
+    ...
+  ]
+}
+```
+
+## TypeScript/JavaScript
+
+The following style guide rules apply for working with TypeScript/JavaScript files.
+
+### TypeScript vs. JavaScript
+
+Whenever possible, write code in TypeScript instead of JavaScript, especially if it's new code. 
+Check out [TYPESCRIPT.md](TYPESCRIPT.md) for help with this process.
+
+### Prefer modern JavaScript/TypeScript syntax
+
+You should prefer modern language features in a lot of cases, e.g.:
+
+* Prefer `class` over `prototype` inheritance
+* Prefer template strings over string concatenation
+
+### Avoid mutability and state
+
+Wherever possible, do not rely on mutable state. This means you should not
+reassign variables, modify object properties, or push values to arrays.
+Instead, create new variables, and shallow copies of objects and arrays:
+
+```js
+// good
+function addBar(foos, foo) {
+  const newFoo = {...foo, name: 'bar'};
+  return [...foos, newFoo];
+}
+
+// bad
+function addBar(foos, foo) {
+  foo.name = 'bar';
+  foos.push(foo);
+}
+```
+
+### Return/throw early from functions
+
+To avoid deep nesting of if-statements, always return a function's value as early
+as possible. And where possible, do any assertions first:
+
+```js
+// good
+function doStuff(val) {
+  if (val > 100) {
+    throw new Error('Too big');
+  }
+
+  if (val < 0) {
+    return false;
+  }
+
+  // ... stuff
+}
+
+// bad
+function doStuff(val) {
+  if (val >= 0) {
+    if (val < 100) {
+      // ... stuff
+    } else {
+      throw new Error('Too big');
+    }
+  } else {
+    return false;
+  }
+}
+```
+
+### Use object destructuring
+
+This helps avoid temporary references and helps prevent typo-related bugs.
+
+```js
+// best
+function fullName({ first, last }) {
+  return `${first} ${last}`;
+}
+
+// good
+function fullName(user) {
+  const { first, last } = user;
+  return `${first} ${last}`;
+}
+
+// bad
+function fullName(user) {
+  const first = user.first;
+  const last = user.last;
+  return `${first} ${last}`;
+}
+```
+
+### Use array destructuring
+
+Directly accessing array values via index should be avoided, but if it is
+necessary, use array destructuring:
+
+```js
+const arr = [1, 2, 3];
+
+// good
+const [first, second] = arr;
+
+// bad
+const first = arr[0];
+const second = arr[1];
+```
+
+### Declare one variable per line, wherever it makes the most sense
+
+This makes it easier to re-order the lines. However, declari variables
+wherever it makes sense (not necessairly on the top of a function).
+
+```js
+// good
+const keys = ['foo', 'bar'];
+const values = [23, 42];
+
+// bad
+const keys = ['foo', 'bar'],
+      values = [23, 42];
+```
+
+### Prefix private class methods with an underscore `JS only`
```

Comment:
I would get rid of this one, or even make it the opposite recommendation. A) it's js only so not really relevant anymore since everyone should be writing in typescript. B) even if it is js, it should eventually be ts, at which point we'll have to rewrite it all to not have underscores.